### PR TITLE
Improved connection error handling for MongoDB interface

### PR DESCRIPTION
**Pull request by penenadpi**
- Fixed error handling in case of non-working Mongo DB instance or any other connection error.
- Added support for disabling persistence manually

### CODE REVIEW

File: src/iac_scan_runner/api.py [hunk 1]
```diff
@@ -202,4 +202,22 @@ async def delete_scan_result(uuid: str) -> JSONResponse:
...
+    try:
+        if(enable == "disable"):
+            scan_runner.persistence_enabled = False              
+        else:
+            scan_runner.persistence_enabled = True
+                         
+        return JSONResponse(status_code=status.HTTP_200_OK, content=f"Persistence enable: {enable}")
+
+    except Exception as e:
+        return JSONResponse(status_code=status.HTTP_400_BAD_REQUEST, content=str(e))
```
Comment (anzoman) [inline src/iac_scan_runner/api.py:223]: Remove and put this logic when the IaC Scan Runner is starting.

File: src/iac_scan_runner/results_persistence.py [hunk 2]
```diff
@@ -3,16 +3,26 @@
...
     
         """
         Initialize new scan result database, collection and client
         """        
-        self.myclient = pymongo.MongoClient("mongodb://localhost:27017/") 
-        self.mydb = self.myclient["scandb"]
-        self.mycol = self.mydb["results"]
+        
+        try:
+            connection_string = os.environ['MONGO_STRING']
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:15]: Rename connection string env var to something like `MONGODB_CONNECTION_STRING`.

File: src/iac_scan_runner/results_persistence.py [hunk 3]
```diff
@@ -3,16 +3,26 @@
...
         """
         Initialize new scan result database, collection and client
         """        
-        self.myclient = pymongo.MongoClient("mongodb://localhost:27017/") 
-        self.mydb = self.myclient["scandb"]
-        self.mycol = self.mydb["results"]
+        
+        try:
+            connection_string = os.environ['MONGO_STRING']  
+            print(connection_string)
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:16]: Remove.

File: src/iac_scan_runner/results_persistence.py [hunk 4]
```diff
@@ -3,16 +3,26 @@
...
-        self.myclient = pymongo.MongoClient("mongodb://localhost:27017/") 
-        self.mydb = self.myclient["scandb"]
-        self.mycol = self.mydb["results"]
+        
+        try:
+            connection_string = os.environ['MONGO_STRING']  
+            print(connection_string)  
+            self.myclient = pymongo.MongoClient(connection_string)
+            self.mydb = self.myclient["scandb"]
+            self.mycol = self.mydb["results"]
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:19]: Check if this can be somehow validated.

**Review by anzoman — CHANGES_REQUESTED** [review]

File: src/iac_scan_runner/api.py (hunk 1 above)
Comment (penenadpi) [reply src/iac_scan_runner/api.py:223]: Done, moved to pre-start checks and used environment variable instead for turning it on/off

File: src/iac_scan_runner/results_persistence.py (hunk 2 above)
Comment (penenadpi) [reply src/iac_scan_runner/results_persistence.py:15]: Done

File: src/iac_scan_runner/results_persistence.py (hunk 3 above)
Comment (penenadpi) [reply src/iac_scan_runner/results_persistence.py:16]: Done

File: src/iac_scan_runner/results_persistence.py (hunk 4 above)
Comment (penenadpi) [reply src/iac_scan_runner/results_persistence.py:19]: Added checks if connection worked and objects are not None

File: src/iac_scan_runner/api.py [hunk 5]
```diff
@@ -167,15 +167,15 @@ async def post_scan(iac: UploadFile = File(..., description='IaC file (zip or ta
...
+async def get_scan_result(uuid: Optional[str]) -> JSONResponse:
     """
     Retrieve a particular scan result (GET method)
     :param uuid: Identifier of a scan record in database
     :return: JSONResponse object (with status code 200 or 400)
     """
     try:
         results_persistence = ResultsPersistence() 
-        if(uuid):           
+        if(uuid and uuid!=""):
```
Comment (anzoman) [inline src/iac_scan_runner/api.py:178]:
```suggestion
        if uuid:           
```

Formatting is needed here, and also empty string is alreafy treated as False.

File: src/iac_scan_runner/api.py [hunk 6]
```diff
@@ -167,15 +167,15 @@ async def post_scan(iac: UploadFile = File(..., description='IaC file (zip or ta
         return JSONResponse(status_code=status.HTTP_400_BAD_REQUEST, content=str(e))
         
 @app.get("/results", summary="Retrieve particular scan result by given uuid", responses={200: {}, 400: {"model": str}})
-async def get_scan_result(uuid: str) -> JSONResponse:
+async def get_scan_result(uuid: Optional[str]) -> JSONResponse:
```
Comment (anzoman) [inline src/iac_scan_runner/api.py:170]: What happens when UUID is not set?

File: src/iac_scan_runner/api.py [hunk 7]
```diff
@@ -167,15 +167,15 @@ async def post_scan(iac: UploadFile = File(..., description='IaC file (zip or ta
...
     """
     Retrieve a particular scan result (GET method)
     :param uuid: Identifier of a scan record in database
```
Comment (anzoman) [inline src/iac_scan_runner/api.py:173]: I would mention database in the API, just `Identifier of saved scan record`.

File: src/iac_scan_runner/results_persistence.py [hunk 8]
```diff
@@ -3,16 +3,37 @@
...
         """
         Initialize new scan result database, collection and client
         """        
-        self.myclient = pymongo.MongoClient("mongodb://localhost:27017/") 
-        self.mydb = self.myclient["scandb"]
-        self.mycol = self.mydb["results"]
+        self.connect_db()
+        
+
+
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:16]: Too many blank lines.

File: src/iac_scan_runner/results_persistence.py [hunk 9]
```diff
@@ -3,16 +3,37 @@
...
+    def connect_db(self):
+        """
+        Initialize new scan result database, collection and client
+        """        
+
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:21]: Remove unnecessary blank line.

File: src/iac_scan_runner/results_persistence.py [hunk 10]
```diff
@@ -3,16 +3,37 @@
...
+        
+
+
+    def connect_db(self):
+        """
+        Initialize new scan result database, collection and client
+        """        
+        
+        try:
+            connection_string = os.environ['MONGODB_CONNECTION_STRING']
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:23]: Validate that `connection_string ` is None if `MONGODB_CONNECTION_STRING` not set.

File: src/iac_scan_runner/results_persistence.py [hunk 11]
```diff
@@ -3,16 +3,37 @@
...
+        """        
+        
+        try:
+            connection_string = os.environ['MONGODB_CONNECTION_STRING']  
+            self.myclient = pymongo.MongoClient(connection_string)
+            self.mydb = self.myclient["scandb"]
+            self.mycol = self.mydb["results"]
+            self.connection_problem = False
+            
+        except Exception as e:
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:29]: Can you narrov that down to a more specific Exception.

File: src/iac_scan_runner/results_persistence.py [hunk 12]
```diff
@@ -21,39 +42,51 @@ def insert_result(self, result: dict):
...
         :param result: Dictionary holding the scan summary
-        """         
-        self.mycol.insert_one(self.parse_json(result))
-
+        """     
+        if(self.connection_problem==True):
+            self.connect_db()
+        if(self.mycol!=None):
+            self.mycol.insert_one(self.parse_json(result))
+
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:50]: Formatting needed.

File: src/iac_scan_runner/results_persistence.py [hunk 13]
```diff
@@ -21,39 +42,51 @@ def insert_result(self, result: dict):
...
-        mydoc = self.mycol.find(myquery)
-        for x in mydoc:
-            return str(x)      
+        if(self.connection_problem==True):
+            self.connect_db()
+        if(self.mycol!=None):                
+            myquery = { "uuid": uuid4 }
+            mydoc = self.mycol.find(myquery)
+            for x in mydoc:
+                return str(x)
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:63]: Formatting needed.

File: src/iac_scan_runner/results_persistence.py [hunk 14]
```diff
@@ -21,39 +42,51 @@ def insert_result(self, result: dict):
...
         """Deletes the scan result with given id from database
         :param uuid4: Identifier of a scan result which is about to be deleted
         """
-        myquery = { "uuid": uuid4 }
-        mydoc = self.mycol.delete_one(myquery)
+        if(self.connection_problem==True):
+            self.connect_db()
+        if(self.mycol!=None):          
+            myquery = { "uuid": uuid4 }
+            mydoc = self.mycol.delete_one(myquery)
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:74]: Formatting needed.

File: src/iac_scan_runner/results_persistence.py [hunk 15]
```diff
@@ -21,39 +42,51 @@ def insert_result(self, result: dict):
...
-            output = output + str(doc)
-        return output	                   
+        if(self.connection_problem==True):
+            self.connect_db()
+        if(self.mycol!=None):        
+            cursor = self.mycol.find({})
+            output = ""
+            for doc in cursor:
+                output = output + str(doc)
+            return output
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:89]: Formatting needed.

File: src/iac_scan_runner/results_persistence.py [hunk 16]
```diff
@@ -79,22 +112,29 @@ def result_age(self, uuid4: str) -> int:
...
-        return self.days_passed(scan_ts)            
+        if(self.connection_problem==True):
+            self.connect_db()
+        if(self.mycol!=None):          
+            myquery = { "uuid": uuid4 }
+            mydoc = self.mycol.find(myquery)
+            for x in mydoc:
+                scan_ts = x["time"]
+                    
+            return self.days_passed(scan_ts)
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:123]: Formatting needed.

File: src/iac_scan_runner/results_persistence.py [hunk 17]
```diff
@@ -79,22 +112,29 @@ def result_age(self, uuid4: str) -> int:
...
+            self.connect_db()
+            
+        if(self.mycol!=None):          
+            cursor = self.mycol.find({})
+            scan_ts = ""
+            for doc in cursor:
+                doc_uuid = doc["uuid"]                
+                age = self.result_age(doc_uuid)
+                if(age>14):
+                    self.delete_result(doc_uuid)
```
Comment (anzoman) [inline src/iac_scan_runner/results_persistence.py:140]: Formatting needed.

File: src/iac_scan_runner/scan_runner.py [hunk 18]
```diff
@@ -45,16 +45,24 @@
...
         self.compatibility_matrix = Compatibility()
-        self.results_summary = ResultsSummary()        
+        self.results_summary = ResultsSummary()
+        self.archive_name = ""   
+        if(os.environ.get("SCAN_PERSISTENCE")=="enabled"):
+            self.persistence_enabled = True
+            print("Enabled persistence")         
+        else:
+            print("Disableed persistence")
+            self.persistence_enabled = False
```
Comment (anzoman) [inline src/iac_scan_runner/scan_runner.py:63]: Formatting needed. Also remove prints.

**Review by anzoman — CHANGES_REQUESTED** [review]

File: src/iac_scan_runner/results_persistence.py (hunk 8 above)
Comment (penenadpi) [reply src/iac_scan_runner/results_persistence.py:16]: Deleted

File: src/iac_scan_runner/results_persistence.py (hunk 9 above)
Comment (penenadpi) [reply src/iac_scan_runner/results_persistence.py:21]: Done

File: src/iac_scan_runner/scan_runner.py (hunk 18 above)
Comment (penenadpi) [reply src/iac_scan_runner/scan_runner.py:63]: Deleted and fixed formatting

File: src/iac_scan_runner/results_persistence.py (hunk 11 above)
Comment (penenadpi) [reply src/iac_scan_runner/results_persistence.py:29]: Gonna put this as TODO, there are many possible exceptions to be considered here

File: src/iac_scan_runner/api.py (hunk 5 above)
Comment (anzoman) [reply src/iac_scan_runner/api.py]: Remove all redundant parentheses.

Comment (anzoman) [reply src/iac_scan_runner/api.py]: Will be resolved later.

**Review by anzoman — APPROVED** [review]
Thanks @penenadpi for your contributions and let's merge this one!

**Pull request merged**